### ISSUE TITLE
unable to load certificate

### ISSUE THREAD

**Issue by waja** [matteocorti/check_ssl_cert#44]
Using check_ssl_cert on Debian stretch (9.x) produces the following problem:

```
$ ./check_ssl_cert -v  -H www.google.com
expect available (/usr/bin/expect)
timeout available (/usr/bin/timeout)
found GNU date with timestamp support: enabling date computations
'/usr/bin/openssl s_client' does not support '-servername': disabling virtual server support
downloading certificate to /tmp
parsing the certificate file
The certificate will expire in 69 day(s)
Certificate will not expire
unable to load certificate
140524407637120:error:0D07207B:asn1 encoding routines:ASN1_get_object:header too long:crypto/asn1/asn1_lib.c:101:
SSL_CERT CRITICAL www.google.com: unable to load certificate
140378701767808:error:0906D06C:PEM routines:PEM_read_bio:no start line:crypto/pem/pem_lib.c:691:Expecting: TRUSTED CERTIFICATE|days=69;;;;
$ dpkg -l | grep openssl
ii  openssl                                      1.1.0d-2                    amd64        Secure Sockets Layer toolkit - cryptographic utility
```

There is a [Debian bugreport](https://bugs.debian.org/855253) about that. Searching the network indicated that other applications [have](https://bugs.debian.org/846319) this issue too.

It seems [dehydrated](https://dehydrated.de) found a [solution](https://github.com/lukas2511/dehydrated/commit/7eca8aec5a6679ce5ca507386687d130cc38ce23) for this, maybe this helps.

Many thanks for looking into this and maintaining check_ssl_cert. Cheers, Jan.

**Comment by matteocorti** [comment]
Dear Jan

Can you please post the output with the `-d` option?

Matteo

**Comment by waja** [comment]
Indeed.

```
./check_ssl_cert -v -d  -H www.google.com
[DBG] ROOT_CA = 
expect available (/usr/bin/expect)
timeout available (/usr/bin/timeout)
[DBG] perl available: /usr/bin/perl
[DBG] date available: /bin/date
found GNU date with timestamp support: enabling date computations
[DBG] check_ssl_version: 1.41.0
[DBG] OpenSSL binary: /usr/bin/openssl
[DBG] OpenSSL version: OpenSSL 1.1.0d  26 Jan 2017 (Library: OpenSSL 1.1.0c  10 Nov 2016)
[DBG] System info: Linux rollercoster02 4.9.0-1-amd64 #1 SMP Debian 4.9.6-3 (2017-01-28) x86_64 GNU/Linux
'/usr/bin/openssl s_client' does not support '-servername': disabling virtual server support
downloading certificate to /tmp
[DBG] executing with timeout (15s): echo 'Q' | /usr/bin/openssl s_client   -connect www.google.com:443  -verify 6     2> /tmp/check_ssl_certDwnX1P 1> /tmp/check_ssl_certC8YRiP
[DBG]   /usr/bin/timeout 15 /bin/sh -c "echo 'Q' | /usr/bin/openssl s_client   -connect www.google.com:443  -verify 6     2> /tmp/check_ssl_certDwnX1P 1> /tmp/check_ssl_certC8YRiP"
[DBG] storing a copy of the retrieved certificate in www.google.com.crt
[DBG] storing a copy of the OpenSSL errors in www.google.com.error
parsing the certificate file
[DBG]  subject=C = US, ST = California, L = Mountain View, O = Google Inc, CN = www.google.com
[DBG] CN         = www.google.com
[DBG] CA_O       = issuer=C = US, O = Google Inc, CN = Google Internet Authority G2
[DBG] CA_CN      = issuer=C = US, O = Google Inc, CN = Google Internet Authority G2
[DBG] SERIAL     = 7F672B96E0AEDA70
[DBG] OCSP_URI   = http://clients1.google.com/ocsp
[DBG] ISSUER_URI = http://pki.google.com/GIAG2.crt
[DBG]     Signature Algorithm: sha256WithRSAEncryption
[DBG] Date computations: GNU
The certificate will expire in 69 day(s)
Certificate will not expire
[DBG] OCSP: fetching issuer certificate http://pki.google.com/GIAG2.crt to /tmp/check_ssl_certxT2gby
[DBG] OCSP: issuer certificate type:  data
[DBG] OCSP: converting issuer certificate from DER to PEM
unable to load certificate
140573843776640:error:0D07207B:asn1 encoding routines:ASN1_get_object:header too long:crypto/asn1/asn1_lib.c:101:
[DBG] OCSP: storing a copy of the retrieved issuer certificate to GIAG2.crt
[DBG] OCSP: host = clients1.google.com
[DBG] openssl ocsp support the -header option
[DBG] executing /usr/bin/openssl ocsp -no_nonce -issuer /tmp/check_ssl_certxT2gby -cert /tmp/check_ssl_certC8YRiP  -url http://clients1.google.com/ocsp  -header HOST clients1.google.com
[DBG] OCSP: response = 
SSL_CERT CRITICAL www.google.com: unable to load certificate
140062676890752:error:0906D06C:PEM routines:PEM_read_bio:no start line:crypto/pem/pem_lib.c:691:Expecting: TRUSTED CERTIFICATE|days=69;;;;
```

**Comment by matteocorti** [comment]
The problem is by the conversion of the issuer certificate

```
 $OPENSSL x509 -inform DER -outform PEM -in "${ISSUER_CERT_TMP}" -out "${ISSUER_CERT_TMP}"
```

This version of OpenSSL does not accept the same file as input and output (which seems a bad idea anyway ...)

**Comment by matteocorti** [comment]
The firs issue is fixe but now

```
/usr/bin/openssl ocsp -no_nonce -issuer GIAG2.crt -cert www.google.com.crt  -url http://clients1.google.com/ocsp  -header HOST clients1.google.com
Missing = in header key=value
ocsp: Use -help for summary.
```

**Comment by matteocorti** [comment]
Seems that `oscp -header` has a new format

```
/usr/bin/openssl ocsp -no_nonce -issuer GIAG2.crt -cert www.google.com.crt  -url http://clients1.google.com/ocsp  -header HOST=clients1.google.com
Response verify OK
www.google.com.crt: good
	This Update: Feb 16 07:07:23 2017 GMT
	Next Update: Feb 23 07:07:23 2017 GMT
```

**Comment by bzed** [comment]
What you need is to check if openssl is >= version 1.1, if so use
```
-header HOST=hostname
```
else - as before
```
-header HOST hostname
```

**Comment by bzed** [comment]
You might want to enable the rm -f trap again.  Guess it got accidentally commented out while debugging. 

Am 16. Februar 2017 18:08:43 MEZ schrieb Matteo Corti <notifications@github.com>:
>Closed #44 via 4bb87dd352f4f479faa35eadb81760eef7173e56.
>
>-- 
>You are receiving this because you commented.
>Reply to this email directly or view it on GitHub:
>https://github.com/matteocorti/check_ssl_cert/issues/44#event-965351803

-- 
Diese Nachricht wurde von meinem Android-Mobiltelefon mit K-9 Mail gesendet.

**Comment by matteocorti** [comment]
Thanks! I just committed the fix.
